### DISCUSSION TITLE
JOIN task taking too long to execute

### DISCUSSION THREAD

**Discussion by shekarpcs** [Netflix/conductor#3436] [General]
We have join task in one of the big workflow which is taking longer to finish even the depedent tasks are completed. 

please help what is causing conductor to execute join task so slow

**Comment by bjpirt** [comment]
We're also seeing this - we've implemented a basic test with a dynamic fork and after all of the sub tasks have completed, the Join always seems to take 31 seconds to complete, despite it having no real work to do.

**Reply by shekarpcs** [reply]
This is caused big issue in production , we had to roll back conductor version upgrade in high tensed situation. 

could some one help on this.  We are stuck in the middle.

**Comment by bjpirt** [comment]
Modifying the property `conductor.app.systemTaskWorkerCallbackDuration` enabled us to reduce this. We've set it to 1s and the JOIN completes almost immediately

**Reply by shekarpcs** [reply]
After setting conductor.app.systemTaskWorkerCallbackDuration=1s   some of the workflows in heavy load staying in running state even after completing all task. surprisingly it is happening with same worklfows with join task

**Comment by shekarpcs** [comment]
Let me try this. Thanks a lot, it really helps.

**Comment by john-larson** [comment]
We are experiencing the same problem. And I have opened a new issue under the bugs section: https://github.com/Netflix/conductor/issues/3472

**Comment by john-larson** [comment]
@shekarpcs @bjpirt  Could you please share the latest version you are using that is free of this bug?

**Comment by bjpirt** [comment]
We've only just started using Conductor so haven't used a version without it - v3.13.2 exhibited the behaviour for us

**Comment by shekarpcs** [comment]
We used 3.4.1 before and it was working fine.

Workflows are not set status to complete in high load also.

We are struggling so much with this
issue. We have no help from community.

**Comment by james-deee** [comment]
Yeah we're seeing the same thing. It sounds like it has to do with this change https://github.com/Netflix/conductor/pull/3284

It's unforunate. We actually dont see this behavor in 3.11.3 either. But, it looks like the performance of this change is an unintended side effect? @apanicker-nflx do you have some insight here?

This is currently blocking us from upgrading Conductor because of the pressure it puts on the underlying DB, and because of how slow these JOINs are finishing.

**Reply by Jozefiel** [reply]
We had the same issue, but we limit the CPU in our compose file to only 2.
After changing to 4 CPUs, JOIN was much faster. 
But previously was fast with 2 CPUs  :(

**Comment by james-deee** [comment]
@apanicker-nflx just wanted to see if you have some insight here on this issue. It's blocking us from upgrading Conductor at the moment due to the performance regression of this change.

**Reply by shekarpcs** [reply]
I do not want you go through the same pain what we went through. .Replying quickly .

We had found following parameter which make conductor to execute tasks based on setting. It is not completing after 3seconds.

conductor.app.systemTaskWorkerCallbackDuration=3s

the default value is 30 seconds. Based on this it was taking 30 seconds to complete join task.

**Comment by james-deee** [comment]
@shekarpcs yeah we already changed it to that. But it A) puts an extreme load on the backing datastore and B) still takes 10s to process in our environment.

**Comment by shekarpcs** [comment]
We added horizontal pods for conductor to process normally. You can do that
to increase performance

**Reply by james-deee** [reply]
@shekarpcs thanks. Yeah that doesn't solve the performance issue, due to load against the backing store though. In fact, if anything that might increase it. Really, this change probably needs to be rolled back, but would be really good to hear from @apanicker-nflx if possible.

**Reply by apanicker-nflx** [reply]
Apologies for the delayed response. The JOIN task execution was made asynchronous to improve the efficiency of workflow evaluations in case of workflows containing forked tasks, especially those workflows that fork into 1000s of tasks with each task taking an arbitrarily long amount of time to execute. During synchronous execution, it was found that a lot of evaluation cycles are wasted on join evaluation when every fork task is updated. The changes made as part of #3284 help to alleviate this. 
However, this does add an overhead in case of short running forked tasks (<1s), where the JOIN could hold up the execution for ~30 seconds. We have a backoff evaluation offset scheme in our fork to account for this. Unfortunately, I am unable to create a PR at this time. Please feel free to create one.
Let me try to summarize the changes below - 

Add a new method to WorkflowSystemTask.java
```java
/**
     * Calculate the offset by which task should be postponed for lazy evaluation on the queue
     *
     * @return the computed offset
     */
    public Optional<Long> getEvaluationOffset(T taskMetadata, long defaultOffset) {
        return Optional.empty();
    }
```

Override this method in Join.java
```java
@Override
    public Optional<Long> getEvaluationOffset(JoinTaskMetadata taskMetadata, long defaultOffset) {
        int index = taskMetadata.getPollCount() > 0 ? taskMetadata.getPollCount() - 1 : 0;
        if (index == 0) {
            return Optional.of(0L);
        }
        return Optional.of(Math.min((long) Math.pow(2, index), defaultOffset));
    }
```

Replace [these lines](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/AsyncSystemTaskExecutor.java#L155-L160) in AsyncSystemTaskExecutor.java with
```java
systemTask.getEvaluationOffset(task, systemTaskCallbackTime)
                        .ifPresentOrElse(task::setCallbackAfterSeconds,
                                () -> task.setCallbackAfterSeconds(systemTaskCallbackTime));
                queueDAO.postpone(
                        queueName,
                        task.getTaskId(),
                        task.getWorkflowPriority(),
                        task.getCallbackAfterSeconds());
```

**Gist of the changes:**
The evaluation offset for a given join task will increase with the number of executions, such that for short running forks, join task evaluation will be almost synchronous (0 second delay) and this will max out at 30 seconds.
```
pollNum || offset 
        0       || 0
        1       || 0
        2       || 2
        3       || 4
        4       || 8
        5       || 16
        6       || 30
        20      || 30
```

Hope this helps.

**Reply by james-deee** [reply]
Thanks @apanicker-nflx. I will attempt this in our version, and see if we get acceptable performance in our environement with this change.

**Reply by james-deee** [reply]
@apanicker-nflx just getting to this, but it looks like your instructions are missing other things. For example, it looks like `JoinTaskMetadata` isnt a class in this project. Also looks like `WorkflowSystemTask` is now a parameterized class in your example? 

Any more guidance would be appreciated.

**Reply by manan164** [reply]
Hi @james-deee , Here is the sample [changes](https://github.com/Netflix/conductor/pull/3594). This is for reference purposes. I have tested this change and it works but I am yet to write the test for that.

**Reply by apanicker-nflx** [reply]
Thanks @manan164 
Apologies @james-deee , the branch I pulled the changes from has other unrelated changes that crept through here. The PR above maps the changes correctly though.

**Reply by shekarpcs** [reply]
@apanicker-nflx @james-deee , are these changes in new version conductor?

We would like to upgrade to new version as soon as fix  is available.

**Comment by VerstraeteBert** [comment]
We're still observing up to 30s delays after upgrading to 3.16.7, even after setting CallbackAfterSeconds to 1s for system tasks. Any idea what could still be going wrong?
